# Read in data
CSV file read in.

CSV file downloaded from [Kaggle](https://www.kaggle.com/hacker-news/hacker-news-posts)

In [1]:
#Read in csv
opened_file = open('HN_posts_year_to_Sep_26_2016.csv')
from csv import reader
read_file = reader(opened_file)
hn_data = list(read_file)
hn_header = hn_data[0]
hn = hn_data[1:]

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))


# Summary of data

The data set has 7 columns and 393,119 rows of data (excl. the header)

The columns are:

|0|1|2|3|4|5|6|
|:---|:---|:---|:---|:---|:---|:---|
|id|title|url|num_points|num_comments|author|created_at|
|id of the post|title of the post|the url of the item being linked to|the number of upvotes the post received|the number of comments the post received|the name of the account that made the post|the date and time the post was made ("mm/dd/yyyy hh:mm")

In [13]:
print("Number of columns: ", len(hn_header), "\n")
print(hn_header)

Number of columns:  7 

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [14]:
print("Number of rows of data: ", len(hn), "\n")

explore_data(hn,0,5)

Number of rows of data:  293119 

['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26']


['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24']


['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19']


['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']


['12578979', 'How the Data Vault Enables the Next-Gen Data Warehouse and Data Lake', 'https://www.talend.com/blog/2016/05/12/talend-and-Â\x93the-data-vaultÂ\x94', '1', '0', 'markgainor1', '9/26/2016 3:14']




# Separate data into three lists: ask_posts, show_posts, and other_posts

The three lists as based on the title starting with "ask_posts" or "show_posts". Where a title does not start with one of these two options, the data is added to the other_posts list of lists.

The three list of lists have the following lengths:
- ask_posts 9,139
- show_posts 1,158
- other_posts 273,822

These three list add up to the total rows of the original dataset, which is 293,119

In [15]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print("Number of rows in ask_posts: ", len(ask_posts))
print("Number of rows in show_posts: ", len(show_posts))
print("Number of rows in other_posts: ", len(other_posts))

print("\n", "The first 5 rows of the ask_posts list of lists:", "\n")
explore_data(ask_posts, 0, 5)

Number of rows in ask_posts:  9139
Number of rows in show_posts:  10158
Number of rows in other_posts:  273822

 The first 5 rows of the ask_posts list of lists: 

['12578908', 'Ask HN: What TLD do you use for local development?', '', '4', '7', 'Sevrene', '9/26/2016 2:53']


['12578522', 'Ask HN: How do you pass on your work when you die?', '', '6', '3', 'PascLeRasc', '9/26/2016 1:17']


['12577908', 'Ask HN: How a DNS problem can be limited to a geographic region?', '', '1', '0', 'kuon', '9/25/2016 22:57']


['12577870', 'Ask HN: Why join a fund when you can be an angel?', '', '1', '3', 'anthony_james', '9/25/2016 22:48']


['12577647', 'Ask HN: Someone uses stock trading as passive income?', '', '5', '2', '00taffe', '9/25/2016 21:50']




# Number of comments for each post type

The average number of comments for each post type is:
- ask_post average number of comments: 10.39
- show_post average number of comments: 4.89

It is evident that the ask_posts receive significantly more comments as the average is approximately double that of the show_posts average

In [17]:
total_ask_comments = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments / len(ask_posts)

print("The average number of comments per ask_post is: ", avg_ask_comments)

The average number of comments per ask_post is:  10.393478498741656


In [18]:
total_show_comments = 0

for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    
avg_show_comments = total_show_comments / len(show_posts)

print("The average number of comments per show_post is: ", avg_show_comments)

The average number of comments per show_post is:  4.886099625910612
